### ISSUE TITLE
Bouncer with Two Guars fails to list the methods in Policy

### ISSUE THREAD

**Issue by GustavoSabel** [adonisjs/bouncer#15]
### Package version

3.1.1

### Describe the bug

Typescript encounters an issue locating the method within a policy when the Authentication process involves multiple Guards associated with distinct models. The issue does not arise if both models possess identical fields. However, introducing an additional field to one model or removing a field from it triggers the problem.

Auth configuration:
```ts
const authConfig = defineConfig({
  default: 'driver',
  guards: {
    account: tokensGuard({
      provider: tokensUserProvider({
        tokens: 'accessTokens',
        model: () => import('#models/account'),
      }),
    }),
    driver: tokensGuard({
      provider: tokensUserProvider({
        tokens: 'accessTokens',
        model: () => import('#models/driver'),
      }),
    }),
  },
})
```

Policy:
```ts
export default class PostPolicy extends BasePolicy {
  async createWithAccount(user: Account) {
    return Promise.resolve(!!user)
  }

  async createWithDriver(user: Driver) {
    return Promise.resolve(!!user)
  }
}
```

Controller:
```ts
export default class PostsController {
  async createWithDriver({ bouncer }: HttpContext) {
    await bouncer.with(PostPolicy).authorize('createWithDriver')
    return 'Posto created with driver'
  }

  async createWithAccount({ bouncer }: HttpContext) {
    // Typescript fails to recognize the method 'createWithAccount' from the policy
    await bouncer.with(PostPolicy).authorize('createWithAccount') 
    return 'Post created with account'
  }
}
```

### Expected Behavior
Expect method `createWithAccount` to be recognized

### Actual Behavior
This
```ts
await bouncer.with(PostPolicy).authorize('createWithAccount')
```
Is showing this typescript error: 
> Argument of type '"createWithAccount"' is not assignable to parameter of type '"createWithDriver"'.ts(2345)

### Reproduction repo

https://github.com/GustavoSabel/issue-bouncer-adonisv6

**Comment by thetutlage** [comment]
This is expected because TypeScript cannot narrow types when the user is a union of `Driver | Account`. 

In plain words, we are asking TypeScript to give us authorization methods that can authorize both `Driver | Account`. 

So here are the two options.

- Either the authorization checks have to work for both the user types. That means, the `createWithAccount(user: Account)` has to be adjusted and written like this `createWithAccount(user: Account | Drive)`
- Other option is to narrow the types in the controller before using the bouncer API's. This is how that will look in practice.

```ts
import { Bouncer } from '@adonisjs/bouncer'

  async createWithDriver({ auth }: HttpContext) {
    const user = auth.user as Driver
    const bouncer = new Bouncer(user)
    await bouncer.with(PostPolicy).authorize('createWithDriver')
    return 'Posto created with driver'
  }

  async createWithAccount({ auth }: HttpContext) {
    const user = auth.user as Account
    const bouncer = new Bouncer(user)

    await bouncer.with(PostPolicy).authorize('createWithAccount') 
    return 'Post created with account'
  }
}
```

That is why I always recommend to have a single user entity within your app and use roles to differentiate between the users.

**Comment by GustavoSabel** [comment]
The problem is that we are moving from V5 to V6, and in V5 we did it in that way, and in V5 that was allowed. That was not a problem for us because we had one middleware for Driver and another for Account, and in each route we use only one of them, so the policy only receives one of them, and we don't need to check in each function of each policy if the user is the correct type. 

In case someone else has the same problem, we found a workaround, at least for now. In `initialize_bouncer_middleware`, we did this change:
```diff
declare module '@adonisjs/core/http' {
  export interface HttpContext {
    bouncer: Bouncer<
-     Exclude<HttpContext['auth']['user'], undefined>,
+     Driver & Account,
      typeof abilities,
      typeof policies
    >
  }
}
```

But maybe in the future, we could change our policies to use some kind of assert to check the type of user, like that

```js
async createWithDriver(user: Driver | Account) {
   assertDriver(user)
   // ...
}
